first pass at autotools system

edhartnett opened the pull request:
**This is now ready to merge.**

I am putting it up to ask questions and so you can see what is happening. I am actually doing all this work on a fork of the GPTL that I have taken.

So far I have moved the code files into new directories include, src/clib, and src/flib. Directory test will hold tests. I have not attempted the fortran code yet - first we will get the C library working.

I have introduced the file config.h. It is created by configure. It needs to be included by every C code file (as I have done), and included before any other header file is included. This file is how the results of the configure script are given to the C pre-processor. Right now only one value is set: HAVE_SLASHPROC.

Take a look at the changes and feel free to add questions or comments to the review. As I continue to work on this, the PR will change to reflect my latest commits.

Fixes #9.

edhartnett committed:
first pass at autotools system

edhartnett committed:
removed old makefile

edhartnett commented on Makefile.am:
```diff
@@ -0,0 +1,30 @@
+## This is a automake file, part of the GPTL package.
+# Copyright 2018, see the COPYRIGHT file for more information.
```
@jmrosinski do we have a copyright notice to put in a COPYRIGHT file?

jmrosinski replied:
Yes there is a file COPYING in the main directory. You can either rename it, and/or feel free to modify and let me have a look.

edhartnett commented on Makefile.am:
```diff
@@ -0,0 +1,30 @@
+	@echo '| If you have not already run "make check", then we strongly  |'
+	@echo '| recommend you do so. It does not take very long.            |'
+	@echo '|                                                             |'
+	@echo '| GPTL is developed and maintained at NOAA.                   |'
```
What should we say instead of "at NOAA" here. More detail would be useful.

jmrosinski replied:
I would say just remove the "is developed" comment entirely, since I've been developing it through my last 5 jobs going all the way back to when I was at NCAR.

edhartnett replied:
It gives the package considerable street cred to cite support by prominent research institutions. How about something like:

GPTL was developed with support from multiple institutions, including NOAA and NCAR. It is maintained by the GSD group at NOAA.

jmrosinski replied:
Yes that's better. Would be good to add ORNL to the lab list since I was there for a couple of years and worked on GPTL there also

edhartnett replied:
OK, done.

edhartnett committed:
removed old makefile.in

edhartnett commented on configure.ac:
```diff
@@ -1,53 +1,96 @@
+dnl 		    [echo "/proc not found: HAVE_SLASHPROC=no"])
+dnl AC_CHECK_LIB(pthread,pthread_mutex_init,[echo "pthreads library found: OK to set PTHREADS=yes"],
+dnl 					[echo "pthreads library not found: PTHREADS=no"])
+AC_DEFINE([HAVE_SLASHPROC], [1], [some comment])
```
@jmrosinski what is HAVE_SLASHPROC? That is, what is a good comment to go with the macro? Right now I have "some comment". ;-)

jedwards4b replied:
https://en.wikipedia.org/wiki/Procfs

edhartnett replied:
Thanks Jim! ;-)

jmrosinski replied:
HAVE_SLASHPROC=yes means that the OS supports the "/proc" pseudo-filesystem. If available, GPTL uses it to discover and print current memory usage by the invoking application. If not available, it backs off to "getrusage()" which is less reliable because on many Linux systems it's a placebo that just returns zeros.

edhartnett committed:
adding includes, moved incorretly located file

edhartnett committed:
starting to add tests

edhartnett commented on test/Makefile.am:
```diff
@@ -0,0 +1,16 @@
+AM_CPPFLAGS = -I$(top_srcdir)/include 
+
+# Test programs to be built.
+#check_PROGRAMS = ???.c
```
@jmrosinski I am having a hard time figuring out what is supposed to happen for make test. Is there a C test program that gets built somewhere?

edhartnett committed:
determining values of existing code defines

edhartnett committed:
enabled rest of existing macros and tests, cleaned up

edhartnett committed:
added C test

edhartnett commented:
I have added the first C test.

edhartnett committed:
fixed comment

edhartnett committed:
fixed comment

edhartnett committed:
developing test

edhartnett committed:
added doxygen

edhartnett commented:
I added doxygen support, so now make builds the doc directory, which contains the html documentation. Looks nice at first glance - it found existing documentation. Jim, have you used doxygen before on this code?

Also now locally have Jenkins runs set up for developer build, MPICH build, test coverage build, and memory sanitizer build.

edhartnett committed:
fixed comment

edhartnett committed:
moved perl script to bin directory

edhartnett committed:
fixed comments

edhartnett committed:
more test development

edhartnett committed:
more test development

jmrosinski commented:
No I've not used doxygen before but am not opposed to learning and using it so I have no objection to you installing it.

You'll have to bear with me I'm familiar with *making* pull requests but not with *accepting* them. I'm guessing now is too early to actually merge your mods into master since you're still working on them, right?

For now I've followed your instructions for Step 1 above. I'll have a look at the changes. I tried running "autoconf" (since there was no "configure" there), and got this:

% autoconf 
configure.ac:24: error: possibly undefined macro: AM_INIT_AUTOMAKE
      If this token and others are legitimate, please use m4_pattern_allow.
      See the Autoconf documentation.

Am I doing things right? If not, are your mods ready to be tested, and if so, please let me know which part(s) are ready for examination.

I'm optimistic about getting the autotools to work. We should be able to get rid of all those "macros.make.*" files, and build a shared library.

jmrosinski commented:
Oops sorry I see your comment at the top saying "don't merge yet", so you can ignore my question about merging.

edhartnett commented:
Jim, in order to build with the new system, you do this:

```
autoreconf -i
./configure
make -j check

```
The autoreconf -i installs all needed bits and pieces for autotools to work. End users will not need to call autoreconf, it's just for developers.

We will indeed get rid of all the macro files, but let's just leave them a bit until we have wider portability working.

edhartnett commented:
Also, before using the new build system, you must have reasonably recent versions of autoconf, automake, and libtool installed.

jmrosinski commented:
OK that worked (I think). "autoreconf -i" complained about "could not find m4 subdirectory", then created the directory and populated it. I ran "autoreconf -i" again and it succeeded without complaint. "./configure" followed by "make" also succeeded, which is good news. I see it created both a libgptl.a and libgptl.so and a bunch of other things I'm not sure what they are.

jmrosinski commented:
"make test" runs a bunch of executables inside of ctests/  Many of the
programs inside this directory get built by "make", then run as part of
"make test" (see the Makefile inside of ctests/). Some of them only get
built but not run. They are there for manual testing.

I have been thinking of redoing the full testing structure for a long time,
but haven't had time to do it. I don't like the way many of these tests
rely on ifdefs to do or ignore various aspects of the tests depending on
how the library was built (e.g. HAVE_PAPI). If there is a standard way of
building testing using autotools, maybe you could just take one or a couple
of the tests that are there and implement them in the "proper" way? I could
then follow your template for the rest.

Note also, many of the tests are inside of ftests/ (Fortran tests). There's
probably more in ftests/ than ctests/ because since Fortran support is just
a layer on top of C, I could kill 2 birds with one stone by running the
test in Fortran.

Jim


On Fri, May 18, 2018 at 7:09 AM, Ed Hartnett <notifications@github.com>
wrote:

> *@edhartnett* commented on this pull request.
> ------------------------------
>
> In test/Makefile.am
> <https://github.com/jmrosinski/GPTL/pull/8#discussion_r189262451>:
>
> > @@ -0,0 +1,16 @@
> +## This is a automake file, part of the GPTL package.
> +# Copyright 2018, see the COPYRIGHT file for more information.
> +
> +# This is the automake file for GPTL source code directories.
> +# Ed Hartnett, 5/18/18
> +
> +AM_CPPFLAGS = -I$(top_srcdir)/include
> +
> +# Test programs to be built.
> +#check_PROGRAMS = ???.c
>
> @jmrosinski <https://github.com/jmrosinski> I am having a hard time
> figuring out what is supposed to happen for make test. Is there a C test
> program that gets built somewhere?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jmrosinski/GPTL/pull/8#pullrequestreview-121411252>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABPN7idVx9W97UJP4ncSHEi-oWAIz0Yzks5tzsgKgaJpZM4UErVF>
> .
>

edhartnett committed:
using existing C tests

edhartnett commented:
OK, I have moved my C test (tst_simple.c) to the ctests directory, and added all the tests there, under a variety of circumstances based on what libraries were available at build time, and what options where selected for configure. I based this on inspection of the current Makefile.

Take a look at the new file ctests/Makefile.am and you will see the conditions in which various tests are run.

As you can see, this allows simplification of the test itself, and we can remove a lot of ifdefs. I have done this for ctests/avail.c. It is now only run if the PAPI library is present, so that does not need to be checked in the test itself.

I have not tested all variants of configurations and library availability. These are the various tests I will add to my jenkins page for GPTL. I suggest you set up jenkins on your machine and start adding configurations as well. Then you can see, for example, the test cove coverage.

edhartnett committed:
changed attribution in makefile.am

edhartnett committed:
dealing with papi

edhartnett committed:
commented out some test in testpapi in order to get it to pass

edhartnett committed:
now run script for parallel testing

edhartnett committed:
handling parallel tests

edhartnett committed:
fixed parallel building

edhartnett committed:
fixed running of pmpi test

edhartnett committed:
look for gettimeofday

edhartnett committed:
further build system development

edhartnett committed:
fixed number of procs

edhartnett committed:
better handling for fortran wrappers

edhartnett committed:
adding ENABLE_PMPI to config.h

edhartnett committed:
fixed configure

edhartnett committed:
more development

edhartnett committed:
now all tests pass

edhartnett committed:
no longer distribute private.h

edhartnett committed:
added support for HAVE_LIBRT

edhartnett committed:
removed unused macro MPI_CONST

edhartnett committed:
added checking for functions MPI_Comm_f2c

edhartnett committed:
took out CONST

edhartnett committed:
added nanotime

edhartnett commented:
@jmrosinski can we get this merged and resolve remaining issues iteratively?

jmrosinski commented:
Yes, too bad we weren't able to do this together yesterday. Are  you
available some time to sit down with me to do this? Since it's going across
repositories I'm not sure how to make it happen. If it's as simple as
pressing a button you probably don't need to be there. Since a number of
local folks are using PGI/MPI I was hoping to address that issue before
putting the code on master, but I guess that can be done afterward.



On Fri, Jun 8, 2018 at 6:03 AM Ed Hartnett <notifications@github.com> wrote:

> @jmrosinski <https://github.com/jmrosinski> can we get this merged and
> resolve remaining issues iteratively?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jmrosinski/GPTL/pull/8#issuecomment-395740213>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABPN7sloF109juPemAY44tQhDNHK0jiVks5t6mgCgaJpZM4UErVF>
> .
>

edhartnett commented:
You should see a button: merge pull request. It will ask you to confirm with another button. That's it.

Let's get these changes merged to master and then proceed with the PGI/MPI work immediately.

Recall that nothing is lost and your old build system is still there in git. You can still use it any time.

edhartnett commented:
Sorry, I seemingly closed this PR by mistake. It's ready to be merged now...

edhartnett commented:
@jmrosinski please merge this. Let me know if you need help.

edhartnett commented:
@jmrosinski what is the status here? We seem to have ground to a halt...

jmrosinski commented:
Sorry for the delay. As you can imagine I've been pretty slammed lately.

OK I hit the "merge" button and it merged the autoconf mods into the main
repo with no complaints. Others can get the previous stuff by checking out
the latest tag.

Can you outline where we go from here? I'm a bit unclear on where the two
of us will be making continuing changes. Presumably it needs to be in the
same repo.

One other thing: Is it possible to run "git difftool" on your (merged from)
repo vs. my (now merged) repo and verify zero diffs?


On Mon, Jun 18, 2018 at 6:03 AM Ed Hartnett <notifications@github.com>
wrote:

> @jmrosinski <https://github.com/jmrosinski> what is the status here? We
> seem to have ground to a halt...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jmrosinski/GPTL/pull/8#issuecomment-398031475>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABPN7ifcRs0GDvcPR3-75OeZwp_zSVCMks5t95b7gaJpZM4UErVF>
> .
>

edhartnett commented:
OK, I have added a few issues, and we already have plenty of others, so we can start working them one at a time.

Yes, I can verify zero diffs from my fork very easily - by creating a pull request from my fork to the repo. The PR will have zero code changes. But you don't have to worry about that, I will keep my fork up to date.

If you want to work on an issue, make a note in the comments (so I know that you are doing it), take a branch, and make your changes. Then put the branch up for a pull request.

I will do the same, but using my fork, since I can't create branches in your repo. But it works the same.
